engines/registry/app/controllers/registry/registry_controller.rb

```diff
@@ -0,0 +1,99 @@
+    before_action :catalog_token_auth, only: [ :catalog ]
+
+    rescue_from Exception, with: :handle_exceptions
+
+    # AuthZ handler
+    # AuthZ will validate which of the requested scope policies are fulfilled
+    # with the current login access and prepare the token to be sent back to the client
+    def authorize
+      token = Registry::AccessToken.new(@client&.account, params['service'], @requested_scopes.map { |s| s.granted(client: @client) }).token
+      render json: { token: token }, status: :ok
+    end
+
+    # Catalog handler
+    # Returns a Distribution Registry HTTP API V2 - compatible repository catalog as defined in
+    # https://distribution.github.io/distribution/spec/api/#listing-repositories
+    def catalog
+      repos = @client.matching_policies.repositories(account_numbers: @client.account_numbers)
+      logger.debug("Returning #{repos.size} repos for client #{@client}")
+
+      response.set_header('Docker-Distribution-Api-Version', REGISTRY_API_VERSION)
+      render json: { repositories: repos }, status: :ok
+    end
+
+    private
+
+    # Support multiple scopes. Podman & Docker handle this differently
+    #   - Podman sends multiple querystrings with the same name.
+    #   - Docker sends space-separated values in only one query string.
+    # This should normalize everything
+    def set_requested_scopes
+      raw_scopes = CGI.parse(request.env['QUERY_STRING']).fetch('scope', [])
+      .join(' ')
+      .split
+      .map(&:presence)
+      .compact
+      .map(&:downcase)
+
+      @requested_scopes = []
+      unless raw_scopes.empty?
+        @requested_scopes = raw_scopes.map do |scope| #.filter_map do |scope|
+          Registry::AccessScope.parse(scope)
+        end
+      end
+      logger.info("Requested scopes: #{@requested_scopes.map(&:to_s)}")
+    end
+
+    # AuthN handler
+    # https://docs.docker.com/registry/spec/auth/jwt/#getting-a-bearer-token
+    def basic_auth
+      # skip authentication if this is not a login request
+      return unless request.authorization
+
+      authenticate_or_request_with_http_basic('SUSE Registry Authentication') do |login, password|
+        begin
+          @client = Registry::AuthenticatedClient.new(login, password)
+        rescue StandardError
+          logger.info _('Could not find system with login \"%{login}\" and password \"%{password}\"') %
+            { login: login, password: password }
+          error = ActionController::TranslatedError.new(N_('Invalid registry credentials'))
+          error.status = :unauthorized
+          raise error
+        end
+
+        true
+      end
+    end
+
+    def catalog_token_auth
+      authenticate_or_request_with_http_token(authorize_api_registry_url, 'authentication required') do |token|
+        begin
+          @client = Registry::CatalogClient.new(token)
+        rescue JWT::DecodeError
+          respond_with_error(message: 'Invalid token', status: :unauthorized) and return
```

Comment:
We cannot simply tell the user that the token is invalid.

Reply:
We are replicating a OCI compliant registry API here and this is the expected response in case the token validation fails.
The client (docker/podman) will just get a new token, then. Usually no user action required.

Reply:
There will be no new token without re-authentication, i.e. we want/need to verify that the instance from which the request is coming from is still the instance that triggered the token generation when the system was registered. When the token expires, i.e. when we remove the system from the auth cache, then the token is expired until a re-authentication process happens. Since neither podman nor docker can send us the necessary data, see https://github.com/containers/podman/discussions/21697, we have to work around the limitation of those tools with our own code. Per agreement today that is going to be `cloudguestregistryauth`

Reply:
That said if the standard only allows for this response then that is the response we have to give and we have to figure out how to reduce the number of tickets generated via some other means